rust-runtime/aws-smithy-client/src/lib.rs

```diff
@@ -84,15 +84,18 @@ pub mod conns {
     /// Return a default HTTPS connector backed by the `hyper_tls` crate.
     ///
     /// It requires a minimum TLS version of 1.2.
-    /// It allows you to connect to both `http` and `https` URLs.
+    /// It allows only https connections. 
+    /// To enable http connections, call `https_only(false)` on the retured NativeTls. 
     pub fn native_tls() -> NativeTls {
         let mut tls = hyper_tls::native_tls::TlsConnector::builder();
         let tls = tls
             .min_protocol_version(Some(hyper_tls::native_tls::Protocol::Tlsv12))
             .build()
             .unwrap_or_else(|e| panic!("Error while creating TLS connector: {}", e));
         let http = hyper::client::HttpConnector::new();
-        hyper_tls::HttpsConnector::from((http, tls.into()))
+        let mut https = hyper_tls::HttpsConnector::from((http, tls.into()));
+        https.https_only(true);
```

Comment:
I assume we'd want the same change in `HTTP_NATIVE_ROOTS` as well, a few lines above.

Reply:
I was thinking about that, but the `hyper_rustls::HttpsConnector` doesn't seem to provide a way for customers to also allow http. So the original concern is still valid here.

Reply:
I'd be quite wary of having diverging behaviour between the two TLS backends we support - that can turn out to be quite surprising for users.

Reply:
If you think that consistency is (more) important in this case, then this PR can be scrapped.

Reply:
My suggestion here is to apply the change to both backends, rather than to `native-tls` only.

Reply:
So it looks like we would need a breaking change for this, either returning the builder from `fn https()` or taking an argument for `allow_http: bool` on it, right?

Reply:
Made a PR for this in [`hyper-rustls` here](https://github.com/rustls/hyper-rustls/pull/191).

Reply:
`rustls` uses a type-state builder, which requires configuration to be supplied in a specific order.
The "step" that you are currently in bleeds through the type signature as well.
Returning a partially configured `rustls` HTTPS connector doesn't look like an option - we need to provide one set of configuration with good defaults and then explain the documentation section how they could configure it otherwise (linking to `rustls`' own docs).